test/e2e/operator/deployment_api.go

```diff
@@ -347,7 +482,147 @@ func startOperator(c *deploy.Cluster, d *deploy.Deployment) {
 		return false
 	}, "3m", "1s").Should(BeTrue(), "increase operator deployment replicas to 1")
 
-	framework.Logf("Ensure operator pod is eady.")
+	framework.Logf("Ensure operator pod is ready.")
 	deploy.WaitForOperator(c, d.Namespace)
 	framework.Logf("Operator is restored!")
 }
+
+func switchDeploymentMode(c *deploy.Cluster, f *framework.Framework, depName string, mode api.DeviceMode) api.Deployment {
+	dsVersion := ""
+	dsName := depName + "-node"
+
+	Eventually(func() error {
+		ds, err := c.GetDaemonSet(dsName, corev1.NamespaceDefault)
+		if err != nil {
+			return err
+		}
+		dsVersion = ds.ObjectMeta.ResourceVersion
+		return nil
+	}, "3m", "1s").ShouldNot(HaveOccurred(), "Get daemonset")
+
+	By(fmt.Sprintf("Switching driver mode to '%s'", mode))
+	deployment := deploy.GetDeploymentCR(f, depName)
+	deployment.Spec.DeviceMode = mode
+	deployment = deploy.UpdateDeploymentCR(f, deployment)
+
+	// Wait till the daemon set get updated
+	Eventually(func() bool {
+		ds, err := c.GetDaemonSet(dsName, corev1.NamespaceDefault)
+		deploy.LogError(err, "Failed to fetch daemon set: %v", err)
+		if err != nil {
+			return false
+		}
+		return ds.ObjectMeta.ResourceVersion != dsVersion
+	}, "3m", "1s").Should(BeTrue(), "Update daemonset")
```

Comment:
This still has the race between "daemon set updated" and "pod restarted".

Better identify the controller pod with `GetAppInstance`, switch the mode, then wait for that pod to be gone with https://github.com/kubernetes/kubernetes/blob/ca7a906ab793c13f3def2d7ff0357ab88bfecf92/test/e2e/framework/pod/wait.go#L380

Reply:
This didn't take into account that changing the driver mode will not restart the controller, right?

But the race still exists. The fix just becomes more complicated, because you will then have to identify all node pods and wait for those to disappear.